configdialog.cpp

```diff
@@ -518,6 +535,78 @@ void PaletteCustomizeDialog::get_and_save_color(int palette_index) {
 	}
 }
 
+std::wstring PaletteCustomizeDialog::export_palettes() {
+	auto all_palettes = m_palette_repo.get_all_custom_palettes();
+	std::wstring palettes_string = L"";
+
+	for (const auto &palette : all_palettes) {
+		auto name = palette.name;
+		auto *data = palette.data;
+
+		palettes_string.append(palette.name + L": ");
+		palettes_string.append(m_palette_repo.serialize(*data));
+		palettes_string.append(L";\r\n");
+	}
+	palettes_string.append(L"~");
+
+	return palettes_string;
+}
+
+void PaletteCustomizeDialog::import_palettes(std::wstring *palettes) {
+	static std::wstring valid_chars = L"qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM0123456789-_!?:";
+	static std::wstring valid_hex_chars = L"0123456789abcdefABCDEF";
+	size_t current_index = palettes->find_first_of(valid_chars.c_str());
+	std::wstring last_successful_palette_name;
+	std::wstring current_name;
+	std::wstring current_data[7];
+	bool graceful_exit = false;
+	try {
+		while (true) {
+			current_name = palettes->substr(current_index, palettes->find_first_of(L':', current_index) - current_index);
+			if (current_name.find_first_not_of(valid_chars + L" ") != std::wstring::npos) {
+				throw ERROR_BAD_ARGUMENTS;
+			}
+			for (int i = 0; i < 7; i++) {
+				current_index = palettes->find_first_of(L'#', current_index) + 1;
+				if (palettes->find_first_of(L';', current_index) - current_index != 6) {
+					throw ERROR_BAD_ARGUMENTS;
+				}
+				current_data[i] = palettes->substr(current_index, 6);
+				if (current_data[i].find_first_not_of(valid_hex_chars) != std::wstring::npos) {
+					throw ERROR_BAD_ARGUMENTS;
+				}
+				current_index = palettes->find_first_of(L';', current_index);
+			}
+			last_successful_palette_name = current_name;
+			if (palettes->find_first_of(valid_chars, current_index) == std::wstring::npos) {
+				if (palettes->find_first_of(L'~', current_index) != std::wstring::npos) {
+					graceful_exit = true;
+				}
+				break;
+			}
+			else {
+				current_index = palettes->find_first_of(valid_chars, current_index);
+			}
+		}
+		if (!graceful_exit) {
+			MessageBox(
+				NULL,
+				(L"A terminating character was not found. One or more palettes may have failed to import. Last successful import: " + last_successful_palette_name).c_str(),
```

Comment:
It looks like the intent of this code is to convert and save as many palettes as possible before running into an error. I argue that this should be an all-or-nothing operation - either every palette converts and is saved, OR none of them are.

My rationale: If a user sees a message saying "whoops, we converted some of your palettes, but not all of them!" they now have to _turn on their brain_ to figure out what to do with that information. Should I delete the ones that converted successfully before and try again? Which ones even converted successfully? Now I have to look at the message box and see which ones came before the last successful import! But shit, I already closed the message box... and crap, I have all these palettes in my list, I can't even remember which ones I was trying to import! Gosh, I feel stupid, computers just aren't for me :(

Meanwhile, if the message box just said "Fuck you," then their only option is to try copy and pasting again or complain to someone. The world is infinitely simpler.